packages/expo-structured-headers/ios/EXStructuredHeaders/EXStructuredHeadersParser.m

```diff
@@ -0,0 +1,534 @@
+      return [outputString copy];
+    } else {
+      [outputString appendFormat:@"%c", nextChar];
+      [self advance];
+    }
+  }
+
+  return [outputString copy];
+}
+
+- (nullable NSNumber *)_parseAnIntegerOrDecimalWithError:(NSError ** _Nullable)error
+{
+  // 4.2.4
+  EXStructuredHeadersParserNumberType type = EXStructuredHeadersParserNumberTypeInteger;
+  NSInteger sign = 1;
+  NSMutableString *inputNumber = [NSMutableString stringWithCapacity:20];
+
+  if ([self compareNextChar:'-']) {
+    [self advance];
+    sign = -1;
+  }
+
+  if (![self hasRemaining]) {
+    if (error) *error = [self errorWithMessage:@"Integer or decimal cannot be empty"];
+    return nil;
+  }
+
+  if (![self isDigit:[self peek]]) {
+    if (error) *error = [self errorWithMessage:@"Integer or decimal must begin with a digit"];
+    return nil;
+  }
+
+  while ([self hasRemaining]) {
+    unichar nextChar = [self consume];
+    if ([self isDigit:nextChar]) {
+      [inputNumber appendFormat:@"%c", nextChar];
+    } else if (type == EXStructuredHeadersParserNumberTypeInteger && nextChar == '.') {
+      if (inputNumber.length > 12) {
+        if (error) *error = [self errorWithMessage:@"Decimal cannot have more than 12 digits before the decimal point"];
+        return nil;
+      }
+      [inputNumber appendFormat:@"%c", nextChar];
+      type = EXStructuredHeadersParserNumberTypeDecimal;
+    } else {
+      [self backout];
+      break;
+    }
+
+    if (type == EXStructuredHeadersParserNumberTypeInteger && inputNumber.length > 15) {
+      if (error) *error = [self errorWithMessage:@"Integer cannot have more than 15 digits"];
+      return nil;
+    } else if (type == EXStructuredHeadersParserNumberTypeDecimal && inputNumber.length > 16) {
+      if (error) *error = [self errorWithMessage:@"Decimal cannot have more than 16 characters"];
+      return nil;
+    }
+  }
+
+  if (type == EXStructuredHeadersParserNumberTypeInteger) {
+    return @(inputNumber.longLongValue * sign);
+  } else {
+    if ([inputNumber hasSuffix:@"."]) {
+      if (error) *error = [self errorWithMessage:@"Decimal cannot end with the character '.'"];
+      return nil;
+    }
+    if ([inputNumber rangeOfString:@"."].location + 3 < inputNumber.length - 1) {
+      if (error) *error = [self errorWithMessage:@"Decimal cannot have more than 3 digits after the decimal point"];
+      return nil;
+    }
+    return @(inputNumber.doubleValue * sign);
+  }
+}
+
+- (nullable NSString *)_parseAStringWithError:(NSError ** _Nullable)error
+{
+  // 4.2.5
+  NSMutableString *outputString = [NSMutableString stringWithCapacity:[self remainingLength]];
+
+  if (![self compareNextChar:'"']) {
+    if (error) *error = [self errorWithMessage:@"String must begin with the character '\"'"];
+    return nil;
+  }
+
+  [self advance];
+  while ([self hasRemaining]) {
+    unichar nextChar = [self consume];
+    if (nextChar == '\\') {
```

Comment:
isn't this 2 characters?

Reply:
Or is this equivalent to `\u005c` in obj c?

Reply:
It's a single character -- it's a unichar, not a string.